### ISSUE TITLE
Export Button

### ISSUE THREAD

**Issue by fzaninotto** [marmelab/react-admin#94]
In the `<List>` view, an `<Export>` component should let the user download the related data:
- with the current filters enabled
- unpaginated (i.e. the whole data)
- in CSV

Just like `<Filters>`, this component should let the developer choose the fields to export:

``` js
const PostExport = (props) => (
    <Export {...props}>
            <TextField source="id" />
            <ReferenceField label="User" source="userId" reference="users">
                <TextField source="name" />
            </ReferenceField>
            <TextField source="title" />
            <TextField source="body" />
    </Export>
);

export const PostList = (props) => (
    <List {...props} export={PostExport}>
        // ...
    </List>
);
```

Not sure if using fields makes sense here, or if we can just pass an object mapping source to formatted data:

``` js
const PostExport = (props) => (
    <Export {...props} fields={{
         Id: { source: 'id' },
         User: { source: 'userId', reference: 'users', referenceSource: 'name' },
         Title: { source: 'title', transform: data => data.toLowercase() },
         Body: { source: 'body' },
    }}>
);
```

cons: it makes us redevelop reference fetching.

**Comment by fzaninotto** [comment]
Elaborating on my previous comment: using `Field` components is a bad idea as they're intended for HTML, and contain HTML tags (e.g. `<TextField>` outputs `<span>foo</span>`). 

Besides, the `Reference` fields are asynchronous, and as such will be hard to include in a string for the user to download - we'll have to know when all the actions have been dealt with.

So we're left with:

- rewrite all the `Field` components as simple functions, minus the HTML, and rewrite the `Reference` field logic to bypass Redux and look more synchronous, or
- use an export function and leave the reference fetching as an exercise to the user

```js
const PostExport = (props) => (
    <Export {...props} render={record => ({
         Id: record.id,
         UserId: record.userId,
         User: ???,
         Title: record.title.toLowercase(),
         Body: record.body,
    })}>
);
```

**Comment by afilp** [comment]
@fzaninotto I really need this a lot! Thanks for thinking about that, was any progress made to that, in 'next' for example?

**Comment by fzaninotto** [comment]
No, we haven't implemented it in `next`. This is still an open issue.

**Comment by fzaninotto** [comment]
Implemented in 2.2, see #2034
